datafusion/functions/src/datetime/date_bin.rs

```diff
@@ -370,11 +421,93 @@ fn date_bin_impl(
         }
     };
 
-    let origin = match origin {
-        ColumnarValue::Scalar(ScalarValue::TimestampNanosecond(Some(v), _)) => *v,
+    let (origin, is_time) = match origin {
+        ColumnarValue::Scalar(ScalarValue::TimestampNanosecond(Some(v), _)) => {
+            (*v, false)
+        }
+        ColumnarValue::Scalar(ScalarValue::Time32Millisecond(Some(v))) => {
+            match stride {
```

Comment:
This `match` block is used 4 times. It could be extracted in a helper function and reused.